src/app/reporting/LoggingManagement.h

```diff
@@ -0,0 +1,245 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *
+ * @brief
+ *   Management of the CHIP Event Logging.
+ *
+ */
+#pragma once
+
+#include "EventLoggingTypes.h"
+#include <core/CHIPCircularTLVBuffer.h>
+#include <messaging/ExchangeMgr.h>
+#include <support/PersistedCounter.h>
+#include <util/basic-types.h>
+
+#define CHIP_CONFIG_EVENT_SIZE_RESERVE 196
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+/**
+ * @brief
+ *   Internal event buffer, built around the chip::TLV::CHIPCircularTLVBuffer
+ */
+
+class CircularEventBuffer : public TLV::CHIPCircularTLVBuffer
+{
+public:
+    // for doxygen, see the CPP file
+    CircularEventBuffer(uint8_t * apBuffer, uint32_t aBufferLength, CircularEventBuffer * apPrev, CircularEventBuffer * apNext);
+
+    CHIP_ERROR GetNextBuffer(chip::TLV::TLVReader & aReader, const uint8_t *& aBufStart, uint32_t & aBufLen) override;
+
+    // for doxygen, see the CPP file
+    bool IsFinalDestinationForPriority(PriorityLevel aPriority) const;
+
+    chip::EventNumber VendEventNumber(void);
+    void RemoveEvent(uint16_t aNumEvents);
+
+    // for doxygen, see the CPP file
+    void AddEvent(timestamp_t aEventTimestamp);
+
+private:
+    friend class LoggingManagement;
+    friend class CircularEventReader;
+
+    CircularEventBuffer * mpPrev; //< A pointer CircularEventBuffer storing events less important events
+    CircularEventBuffer * mpNext; //< A pointer CircularEventBuffer storing events more important events
+
+    PriorityLevel mPriority; //< The buffer is the final bucket for events of this priority.  Events of lesser priority are
+                             //< dropped when they get bumped out of this buffer
+
+    chip::EventNumber mFirstEventNumber; //< First event Number stored in the logging subsystem for this priority
+    chip::EventNumber mLastEventNumber;  //< Last event Number vended for this priority
+
+    timestamp_t mFirstEventTimestamp; //< The timestamp of the first event in this buffer
+    timestamp_t mLastEventTimestamp;  //< The timestamp of the last event in this buffer
+
+    // The counter we're going to actually use.
+    chip::MonotonicallyIncreasingCounter * mpEventNumberCounter;
+
+    // The backup counter to use if no counter is provided for us.
+    chip::MonotonicallyIncreasingCounter mNonPersistedCounter;
+};
+
+/**
+ * @brief
+ *   A TLVReader backed by CircularEventBuffer
+ */
+class CircularEventReader : public chip::TLV::TLVReader
+{
+    friend class CircularEventBuffer;
```

Comment:
why is the friend class required here? Are there private/protected member accesses required? If so, why?

Reply:
no needed, just remove. thanks for catching up this.